index.js

```diff
@@ -1,39 +1,56 @@
 import {PassThrough as PassThroughStream} from 'node:stream';
+import {finished} from 'node:stream/promises';
 
 export default function mergeStreams(streams) {
 	if (!Array.isArray(streams)) {
 		throw new TypeError(`Expected an array, got \`${typeof streams}\`.`);
 	}
 
-	const passThroughStream = new PassThroughStream({objectMode: true});
+	const invalidStream = streams.find(stream => typeof stream?.pipe !== 'function');
+	if (invalidStream !== undefined) {
+		throw new TypeError(`Expected a readable stream, got: \`${typeof invalidStream}\`.`);
+	}
+
+	const objectMode = streams.some(({readableObjectMode}) => readableObjectMode);
+	const readableHighWaterMark = getReadableHighWaterMark(streams, objectMode);
+	const passThroughStream = new PassThroughStream({objectMode, writableHighWaterMark: 0, readableHighWaterMark});
 	passThroughStream.setMaxListeners(Number.POSITIVE_INFINITY);
 
 	if (streams.length === 0) {
 		passThroughStream.end();
 		return passThroughStream;
 	}
 
-	let activeStreams = streams.length;
-
 	for (const stream of streams) {
-		if (!(typeof stream?.pipe === 'function')) {
-			throw new TypeError(`Expected a stream, got: \`${typeof stream}\`.`);
-		}
-
 		stream.pipe(passThroughStream, {end: false});
+	}
 
-		stream.on('end', () => {
-			activeStreams--;
+	pipeStream(passThroughStream, streams);
 
-			if (activeStreams === 0) {
-				passThroughStream.end();
-			}
-		});
+	return passThroughStream;
+}
 
-		stream.on('error', error => {
-			passThroughStream.emit('error', error);
-		});
+const getReadableHighWaterMark = (streams, objectMode) => {
+	if (streams.length === 0) {
+		return 0;
 	}
 
-	return passThroughStream;
-}
+	const highWaterMarks = streams
+		.filter(({readableObjectMode}) => readableObjectMode === objectMode)
+		.map(({readableHighWaterMark}) => readableHighWaterMark);
+	return Math.max(...highWaterMarks);
+};
+
+const pipeStream = async (passThroughStream, streams) => {
+	try {
+		await Promise.all(streams.map(stream => finished(stream, {cleanup: true, readable: true, writable: false})));
+		passThroughStream.end();
+	} catch (error) {
+		// This is the error thrown by `finished()` on `stream.destroy()`
```

Comment:
https://github.com/nodejs/node/blob/e57a32a03a9682857ed83c38447291b3e83b983e/lib/internal/streams/end-of-stream.js#L154